### ISSUE TITLE
Consider Alternatives for MapN Combinators

### ISSUE THREAD

**Issue by adamgfraser** [zio/zio#2241]
Currently we provide `MapN` combinators through implicit syntax so we could do this:

```scala
import zio.syntax._

(zio1, zio2, zio3, zio4).map4(f)
```

This is nice but requires importing the syntax, which goes against our general philosophy of providing concrete methods on concrete types and not requiring the user to import implicits.

I wonder if we would be better off with something like:

```scala
zio1.map4(zio2, zio3, zio4)(f)
```

or:

```scala
ZIO.map4(zio1, zio2, zio3, zio4)
```

We could also potentially use some kind of builder syntax like in ZIO config:

```scala
(zio1 |@| zio2 |@| zio3 |@| zio4)(f)
```

Open to feedback on which style people like more but I think it would be nice to provide a convenient way to combine multiple effects like this that we could use across Core, Stream, and Test, potentially also having `zipN` variants for collection like types.

**Comment by saraiva132** [comment]
I like having the method in the companion object a lot `ZIO.map4(zio1, zio2, zio3, zio4)` with `(zio1 |@| zio2 |@| zio3 |@| zio4)(f)` as an alternative

**Comment by saraiva132** [comment]
My question @adamgfraser. Do you know why syntax was added in the first place. And if deprecating syntax is an option for mapN combinators, what about the rest?

**Comment by laurynasl-wix** [comment]
How about the following: 

```scala
final case class Ops2[R, E, A1, A2](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2]) {
  def tupled: ZIO[R, E, (A1, A2)] = ???
  def mapN[O](f: (A1, A2) => O): ZIO[R, E, O] = ???
}

final case class ParOps2[R, E, A1, A2](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2]) {
  def tupled: ZIO[R, E, (A1, A2)] = ???
  def mapN[O](f: (A1, A2) => O): ZIO[R, E, O] = ???
}

final case class Ops3[R, E, A1, A2, A3](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2], zio3: ZIO[R, E, A3]) {
  def tupled: ZIO[R, E, (A1, A2, A3)] = ???
  def mapN[O](f: (A1, A2, A3) => O): ZIO[R, E, O] = ???
}

final case class ParOps3[R, E, A1, A2, A3](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2], zio3: ZIO[R, E, A3]) {
  def tupled: ZIO[R, E, (A1, A2, A3)] = ???
  def mapN[O](f: (A1, A2, A3) => O): ZIO[R, E, O] = ???
}

def many[R, E, A1, A2](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2]) = Ops2(zio1, zio2)
def many[R, E, A1, A2, A3](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2], zio3: ZIO[R, E, A3]) = Ops3(zio1, zio2, zio3)

def manyPar[R, E, A1, A2](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2]) = ParOps2(zio1, zio2)
def manyPar[R, E, A1, A2, A3](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2], zio3: ZIO[R, E, A3]) = ParOps3(zio1, zio2, zio3)
```

This could be used as follows:
```scala
ZIO.many(z1, z2).tupled
ZIO.many(z1, z2, z3).tupled
ZIO.many(z1, z2).mapN(_ + _)
ZIO.many(z1, z2, z3).mapN(_ + _ + _)

ZIO.manyPar(z1, z2).tupled
ZIO.manyPar(z1, z2, z3).tupled
ZIO.manyPar(z1, z2).mapN(_ + _)
ZIO.manyPar(z1, z2, z3).mapN(_ + _ + _)
```

Now if you give the `many` method something that's not a ZIO, it will complain that it can't find an overloaded method. We can help here by also defining version with fixed arity, so that the compiler knows which one we wanted and can tell us which argument is bad.

```scala
  def many2[R, E, A1, A2](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2]) = Ops2(zio1, zio2)
  def many3[R, E, A1, A2, A3](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2], zio3: ZIO[R, E, A3]) = Ops3(zio1, zio2, zio3)

  def manyPar2[R, E, A1, A2](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2]) = ParOps2(zio1, zio2)
  def manyPar3[R, E, A1, A2, A3](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2], zio3: ZIO[R, E, A3]) = ParOps3(zio1, zio2, zio3)
```

I know this is a bit unorthodox, but I'd still love to get some feedback on this pattern.

**Comment by laurynasl-wix** [comment]
Alternatively, the decition if it's parallel or sequential can be move inside `Ops`

```
final case class Ops2[R, E, A1, A2](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2]) {
  def tupled: ZIO[R, E, (A1, A2)] = ???
  def mapN[O](f: (A1, A2) => O): ZIO[R, E, O] = ???
  def parTupled: ZIO[R, E, (A1, A2)] = ???
  def parMapN[O](f: (A1, A2) => O): ZIO[R, E, O] = ???
}

final case class Ops3[R, E, A1, A2, A3](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2], zio3: ZIO[R, E, A3]) {
  def tupled: ZIO[R, E, (A1, A2, A3)] = ???
  def mapN[O](f: (A1, A2, A3) => O): ZIO[R, E, O] = ???
  def parTupled: ZIO[R, E, (A1, A2, A3)] = ???
  def parMapN[O](f: (A1, A2, A3) => O): ZIO[R, E, O] = ???
}

def many[R, E, A1, A2](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2]) = Ops2(zio1, zio2)
def many[R, E, A1, A2, A3](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2], zio3: ZIO[R, E, A3]) = Ops3(zio1, zio2, zio3)

def many2[R, E, A1, A2](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2]) = Ops2(zio1, zio2)
def many3[R, E, A1, A2, A3](zio1: ZIO[R, E, A1], zio2: ZIO[R, E, A2], zio3: ZIO[R, E, A3]) = Ops3(zio1, zio2, zio3)

ZIO.many(z1, z2).tupled
ZIO.many(z1, z2).mapN(_ + _)
ZIO.many(z1, z2).parTupled
ZIO.many(z1, z2).parMapN(_ + _)
ZIO.many(z1, z2, z3).tupled
ZIO.many(z1, z2, z3).mapN(_ + _ + _)
ZIO.many(z1, z2, z3).parTupled
ZIO.many(z1, z2, z3).parMapN(_ + _ + _)
```

**Comment by adamgfraser** [comment]
It looks like we're just going to make these methods on the companion object. See #2361.

@laurynasl-wix What you're describing is an example of the partial application trick. We use it a decent amount (see, for example, the implementation of `checkSome` in ZIO Test. I find it normally makes code more complicated than code written as normal methods so we typically only use it when there is a "problem" where are trying to solve, typically around type inference, avoiding ambiguous method signatures, or not having to specify all the type parameters when some must be specified but others can be inferred.

**Comment by laurynasl-wix** [comment]
@adamgfraser Yes, it is very similar, and that's where the inspiration comes from. Now that I looked at the PR, my suggestion looks like an explicit way to summon the implicit `TupleN` classes. It would still be nice to have a way to summon them explicitly, but that can easily be done in userland.

**Comment by adamgfraser** [comment]
@laurynasl-wix Right. The idea of moving it to the companion object is there isn't any implicit you have to summon. It is just `ZIO.mapN(zio1, zio2, zio3, zio4)(f)`.

**Comment by laurynasl-wix** [comment]
Cool, thanks!

**Comment by adamgfraser** [comment]
Closed by #2361.
